include/RAJA/pattern/forall.hpp

```diff
@@ -399,25 +487,34 @@ forall(ExecutionPolicy&& p, Container&& c, LoopBody&& loop_body)
 
   util::callPreLaunchPlugins(context);
 
-  wrap::forall(std::forward<ExecutionPolicy>(p),
-               std::forward<Container>(c),
-               body);
+  resources::EventProxy<Res> e =  wrap::forall(r,
+                                          std::forward<ExecutionPolicy>(p),
+                                          std::forward<Container>(c),
+                                          std::move(body));
 
   util::callPostLaunchPlugins(context);
+  return e;
 }
 
 }  // end inline namespace policy_by_value_interface
 
 
-/*!
+ /*!
  * \brief Conversion from template-based policy to value-based policy for forall
  *
  * this reduces implementation overhead and perfectly forwards all arguments
  */
-template <typename ExecutionPolicy, typename... Args>
-RAJA_INLINE void forall(Args&&... args)
+template <typename ExecutionPolicy, typename... Args, 
+          typename Res = typename resources::get_resource<ExecutionPolicy>::type >
+RAJA_INLINE resources::EventProxy<Res> forall(Args&&... args)
 {
-  policy_by_value_interface::forall(ExecutionPolicy(), std::forward<Args>(args)...);
+  Res r = Res::get_default();
```

Comment:
Should this get the default resource for ExecutionPolicy.

Reply:
The resource type is deduced on the execution policy in the template definition above.